File: 400/447 - Number of Boomerangs.java
import java.util.HashMap;

public class Solution {
    public int numberOfBoomerangs(int[][] points) {
        int i, j, ret = 0, t;
        long x1, y1, dx, dy, val;
        HashMap<Long, Integer> hm = new HashMap<Long, Integer>();
        
        for (i = 0; i < points.length; ++i) {
            x1 = points[i][0];
            y1 = points[i][1];
            hm.clear();
            
            for (j = 0; j < points.length; ++j) {
                dx = points[j][0] - x1;
                dy = points[j][1] - y1;
                val = dx * dx + dy * dy;
                
                if (hm.containsKey(val))
                    hm.replace(val, hm.get(val) + 1);
                else
                    hm.put(val, 1);
            }
            
            for (HashMap.Entry<Long, Integer> p : hm.entrySet()) {
                t = p.getValue();
                ret += t * (t - 1);
            }
        }
        
        return ret;
    }
}

File: csharp/447.number-of-boomerangs.cs
/*
 * @lc app=leetcode id=447 lang=csharp
 *
 * [447] Number of Boomerangs
 */

using System.Collections.Generic;

public class Solution {
    public int NumberOfBoomerangs(int[][] points) {
        var dict = new Dictionary<int, int>();
        int ret = 0;
        foreach (int[] curr in points) {
            int x = curr[0], y = curr[1];
            foreach (int[] point in points) {
                int dx = x - point[0], dy = y - point[1], d = dx * dx + dy * dy;
                dict[d] = 1 + (dict.ContainsKey(d) ? dict[d] : 0);
            }
            foreach (int val in dict.Values) {
                ret += val * (val - 1);
            }
            dict.Clear();
        }
        return ret;
    }
}

